File: java/Carrinho.java
import java.util.ArrayList;
import java.util.List;

public class Carrinho {

    private List<Produto> produtos;

    public Carrinho() {

        produtos =
                new ArrayList<>();
    }

    public void adicionarProduto(
            Produto produto
    ) {

        if (
                produto.getEstoque() > 0
        ) {

            produtos.add(
                    produto
            );

            produto.reduzirEstoque();

            System.out.println(
                    "Produto adicionado: "
                            + produto.getNome()
            );

        } else {

            System.out.println(
                    "Produto indisponível."
            );
        }
    }

    public void removerProduto(
            Produto produto
    ) {

        if (
                produtos.remove(
                        produto
                )
        ) {

            System.out.println(
                    "Produto removido: "
                            + produto.getNome()
            );

        } else {

            System.out.println(
                    "Produto não encontrado no carrinho."
            );
        }
    }

    public double calcularTotal() {

        double total = 0;

        for (
                Produto produto :
                produtos
        ) {

            total +=
                    produto.getPreco();
        }

        return total;
    }

    public void mostrarCarrinho() {

        System.out.println(
                "\nCarrinho:"
        );

        for (
                Produto produto :
                produtos
        ) {

            System.out.println(
                    "- " +
                    produto.getNome()
            );
        }

        System.out.println(
                "Total: R$ " +
                String.format(
                        "%.2f",
                        calcularTotal()
                )
        );
    }
}

File: backend/models/Carrinho.php
<?php

require_once 'Produto.php';

class Carrinho
{
    private array $produtos = [];

    public function adicionarProduto(
        Produto $produto
    ): bool
    {
        if ($produto->getEstoque() > 0) {

            $this->produtos[] = $produto;

            $produto->reduzirEstoque();

            return true;
        }

        return false;
    }

    public function removerProduto(
        Produto $produto
    ): bool
    {
        foreach ($this->produtos as $key => $item) {

            if (
                $item->getId()
                ===
                $produto->getId()
            ) {

                unset(
                    $this->produtos[$key]
                );

                return true;
            }
        }

        return false;
    }

    public function calcularTotal(): float
    {
        $total = 0;

        foreach (
            $this->produtos
            as $produto
        ) {

            $total +=
                $produto->getPreco();
        }

        return $total;
    }

    public function getProdutos(): array
    {
        return $this->produtos;
    }
}